StringStore/Vocab dev docs

polm opened the pull request:
Things to check:

1. The mysterious vocab members
2. How to make table of contents? Is it autogenerated?
3. Anything I missed / needs more detail?

<!--- Provide a general summary of your changes in the title. -->


### Types of change
<!-- What type of change does your PR cover? Is it a bug fix, an enhancement
or new feature, or a change to the documentation? -->

docs update

## Checklist
<!--- Before you submit the PR, go over this checklist and make sure you can
tick off all the boxes. [] -> [x] -->
- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [ ] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

adrianeboyd commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,212 @@
+Some members of the Vocab are more mysterious. 
+
+- `get_noun_chunks`: This is maybe here for serialization reasons?
+- `cfg`: This is a dict that just stores `oov_prob`, which isn't used (???)
```
I think it should only use `oov_prob` from the lookups table `lexeme_settings`. Possibly it would be better if the default in `Lexeme.prob` were the one from init rather than the hard-coded `-20`. (This is a bad hard-coded default.) I think I wouldn't add `oov_prob` at this point if it weren't already there, but since it is, it makes sense to use it here.

adrianeboyd commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,212 @@
+Most of these are straightforward. `shape` can technically be customized,
+though that's not usually done. `norm` is more likely to be customized, and
+spaCy doesn't prescribe any particular type of normalization.
+
```
These can all be customized in the language defaults (even `lower`). In particular, the prefix and suffix lengths are probably good for English and not optimal for many other languages, but currently there's no much language-specific customization.

polm commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,212 @@
+
+Hashes are 64-bit and are computed using murmurhash on UTF-8 bytes. There is no
+mechanism for detecting and avoiding collisions; to date a collision has never
+been observed.
```
OK, I'll soften the language there a little.

polm commented:
I think I addressed all comments and this is basically ready. Some of the maybe-unused things mentioned at the bottom are still being figured out but that can be updated as those fields are dealt with.

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+issue. (Note that this is also different from using `sys.intern` to intern
+Python strings, which does not guarantee they won't be memory collected later.)
+
+The way strings are stored in the `StringStore` is weird, but is well isolated
+from outside code. A short explanation is that very short strings are stored
+in a fixed-length buffer (one per string, not a shared buffer), while longer
```
```suggestion
Strings are stored in the `StringStore` in a peculiar way: short strings are stored
in a fixed-length buffer (one per string, not a shared buffer), while longer
```

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+though it is not a subclass of anything in particular.
+
+The main functionality of the `StringStore` is that `__getitem__` converts
+hashes into strings or strings into hashes. The details of this are very messy
+but generally not a problem in practice.
```
This sentence feels strange to me. The dev docs should explain what exactly is going on, or point to the code / other resources if need be. Using terms like "generally" and "not a problem in practice" makes the reader feel as if they're still missing out on those details. Similarly, "very messy" doesn't really help - the dev docs should make clear what the peculiarities are and why they were implemented in this fashion (often for performance reasons) or kept in the code as such (e.g. for historical reasons / backwards compatibility)

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+try:
+  # this won't work
+  ss[hashval]
+except:
+  pass
```
```suggestion
except KeyError:
  print(f"key {hashval} unknown in the StringStore.")
```
I would also format this code block with 4 spaces indentation instead of 2 to mimic the normal `black` formatting we use throughout the code base.

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+string types to manage compatability with bytes and strings in Python 2 and 3:
+`str`, `unicode`, `bytes`, and `basestring`. 
+
+Here is how the types work in Python 3:
+
+- `str` and `unicode` are the same
+- `bytes` is just `bytes`
+- `basestring` is the union of `str` and `unicode`, so in Py3 all are equivalent
+
+`basestring` and `unicode` are still used in several places in spaCy code,
+presumably as a legacy of Python 2 support. In Python 3 there should be no
+difference between these types and `str`.
```
This can be updated after the other PR is merged

polm replied:
Since it's been merged I'll just remove this section.

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+While you can ignore the `StringStore` in many cases, there are situations where
+you should make use of it to avoid errors. 
+
+Any time you introduce a string that may be set on a Doc field that has a hash,
```
```suggestion
Any time you introduce a string that may be set on a `Doc` field that has a hash,
```

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+labels in components, but there are some other cases:
+
+- syntax iterators, mainly `get_noun_chunks`
+- external data used in components, like the KnowledgeBase in EntityLinking
```
```suggestion
- external data used in components, like the `KnowledgeBase` in the `entity_linker`
```

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+to manage `Lexeme`s, which are structs that contain information about a token
+that depends only on its surface form, without context. `Lexeme`s store much of
+the data associated with `Token`s. As a side effect of this the `Vocab` also
+managed the `StringStore` for a pipeline and a grab-bag of other data.
```
```suggestion
manages the `StringStore` for a pipeline and a grab-bag of other data.
```

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+- `has_letters`: bool (default `True`, `False` only for CJK for now)
+
+Currently these are not used much - the main use is that `direction` is used in
+visualizers, though `rtl` doesn't quite work. In the future they could be used
```
Explain "doesn't quite work" ;-)

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+
+- `get_noun_chunks`
+- `cfg`: This is a dict that just stores `oov_prob` (hardcoded to `-20`)
+- `data_dir`: This seems to be serialization related but presently unused.
```
Let's discuss the `data_dir` property internally and either remove it from the code base or document it more clearly here.

adrianeboyd commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+strings considered "core" in spaCy. This was critical in v1, which hasn't
+introduced hashing yet. Since v2 it's important for items in `spacy.attrs`,
+especially lexeme flags, but is otherwise only maintained for backwards
+compatability.
```
```suggestion
compatibility.
```

svlandeg commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,220 @@
+
+### Other Vocab Members
+
+Some members of the Vocab have less clear purposes. The Vocab is kind of the
```
Again, this is confusing as a reader. What do you mean by "less clear purposes"? If the next sentence explains it, I would just remove the first sentence / rephrase this paragraph slightly differently.

svlandeg reviewed (commented):
Thanks for taking this on! It's definitely not an easy task to write up the dev docs for something like this. I did have some various comments, cf below. Some are definitely nitpicks ;-)

Also to reply to one of your questions:
> How to make table of contents? Is it autogenerated?

I did this manually. You can look at the preview of the `md` file and click on the subheadings to see what the URL will look like, to easily generate the links.

polm commented:
Thanks for the feedback, I'll go over it!

About the table of contents, if it's manually generated, do we need to include it? Github automatically generates a navigable version based on headings. 

https://user-images.githubusercontent.com/286278/133203795-f3504475-0a7f-4cb1-97ac-b49215afeefb.mp4

polm commented on extra/DEVELOPER_DOCS/StringStore-Vocab.md:
```diff
@@ -0,0 +1,216 @@
+You can call `strings["mystring"]` with a string the `StringStore` has never seen
+before and it will return a hash. But in order to do the reverse operation, you
+need to call `strings.add("mystring")` first. Without a call to `add` the
+string will not be interned.
```
This and the preceding paragraph now feel a bit redundant with the code sample. Should I move the code sample down here?

svlandeg replied:
Yes I think it makes sense to have the explanation first, then the code sample.

svlandeg reviewed (approved):
Apart from the one edit you suggested, I think this is good to merge! We can always edit more later, but I think this is a great start. Feel free to merge whenever you're ready :-)
